File: backend/r2.py
import json
from typing import Any, Dict, List

import boto3
from botocore.config import Config

from .settings import Settings


def _get_r2_client(settings: Settings):
    # Build a boto3 client for Cloudflare R2.
    return boto3.client(
        "s3",
        endpoint_url=settings.R2_ENDPOINT,
        aws_access_key_id=settings.R2_ACCESS_KEY_ID,
        aws_secret_access_key=settings.R2_SECRET_ACCESS_KEY,
        config=Config(signature_version="s3v4"),
        region_name="auto",
    )


def load_manifest(settings: Settings, prefix: str) -> Dict[str, Any]:
    # Download and parse the job manifest from R2.
    client = _get_r2_client(settings)
    key = f"{prefix.rstrip('/')}/manifest.json"
    try:
        response = client.get_object(Bucket=settings.R2_BUCKET, Key=key)
        body = response["Body"].read().decode("utf-8")
        return json.loads(body)
    except Exception as exc:
        raise RuntimeError(f"Failed to load manifest {key}: {exc}") from exc


def sign_clip_urls(settings: Settings, prefix: str, clip_files: List[str]) -> List[str]:
    # Generate signed download URLs for clip files.
    client = _get_r2_client(settings)
    urls: List[str] = []
    for clip_file in clip_files:
        key = f"{prefix.rstrip('/')}/clips/{clip_file}"
        try:
            url = client.generate_presigned_url(
                "get_object",
                Params={"Bucket": settings.R2_BUCKET, "Key": key},
                ExpiresIn=3600,
            )
        except Exception as exc:
            raise RuntimeError(f"Failed to sign URL for {key}: {exc}") from exc
        urls.append(url)
    return urls

File: backend/src/r2.ts
import { GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { Config } from "./config.js";

// Builds an S3 client for Cloudflare R2.
const createClient = (config: Config) => {
  return new S3Client({
    region: "auto",
    endpoint: config.r2Endpoint,
    credentials: {
      accessKeyId: config.r2AccessKeyId,
      secretAccessKey: config.r2SecretAccessKey,
    },
  });
};

// Converts a stream body to a UTF-8 string.
const streamToString = async (body: any): Promise<string> => {
  const chunks: Uint8Array[] = [];
  for await (const chunk of body) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks).toString("utf-8");
};

// Loads and parses the manifest JSON from R2.
export const loadManifest = async (config: Config, prefix: string) => {
  const client = createClient(config);
  const key = `${prefix.replace(/\/+$/, "")}/manifest.json`;
  const command = new GetObjectCommand({
    Bucket: config.r2Bucket,
    Key: key,
  });
  try {
    const response = await client.send(command);
    if (!response.Body) {
      throw new Error("Missing body");
    }
    const body = await streamToString(response.Body);
    return JSON.parse(body) as Record<string, unknown>;
  } catch (err) {
    throw new Error(`Failed to load manifest ${key}: ${String(err)}`);
  }
};

// Creates presigned URLs for clip files.
export const signClipUrls = async (
  config: Config,
  prefix: string,
  clipFiles: string[]
) => {
  const client = createClient(config);
  const urls: string[] = [];
  for (const clipFile of clipFiles) {
    const key = `${prefix.replace(/\/+$/, "")}/${clipFile}`;
    try {
      const url = await getSignedUrl(
        client,
        new GetObjectCommand({
          Bucket: config.r2Bucket,
          Key: key,
        }),
        { expiresIn: 3600 }
      );
      urls.push(url);
    } catch (err) {
      throw new Error(`Failed to sign URL for ${key}: ${String(err)}`);
    }
  }
  return urls;
};
